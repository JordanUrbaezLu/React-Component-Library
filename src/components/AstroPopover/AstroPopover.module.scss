.astroPopoverContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
}

.astroPopover {
  color: rgb(223, 223, 223);
  position: relative;
  text-align: center;
  width: fit-content;
  max-width: 200px;
  background-color: rgb(60, 60, 60);
  border-radius: 4px;
  box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
  padding: 8px 10px;
  text-overflow: wrap;
  word-wrap: break-word;
  margin-top: 4px;
  z-index: 999;
  &.nubbin {
    &::before {
      content: "";
      width: 10px;
      height: 10px;
      z-index: 1000;
      background-color: rgb(60, 60, 60);
      position: absolute;
      transform: translate(-50%, -50%) rotate(45deg);
    }
    &.bottom {
      margin-top: 11px;
    }
  }
}

.bottom {
  &::before {
    top: 0;
    left: 50%;
  }
}

.top {
  &::before {
    top: 100%;
    left: 50%;
  }
}
