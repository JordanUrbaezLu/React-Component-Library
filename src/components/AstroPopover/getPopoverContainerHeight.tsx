import * as React from "react";

export const getPopoverContainerHeight = ({
  ref,
}: {
  ref: React.RefObject<HTMLDivElement>;
}) => {
  return ref.current?.getBoundingClientRect().height;
};
