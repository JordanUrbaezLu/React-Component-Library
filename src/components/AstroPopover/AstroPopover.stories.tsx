import * as React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import AstroPopover from "./AstroPopover";
import AstroButton from "../AstroButton/AstroButton";

export default {
  title: "AstroPopover",
  componet: AstroPopover,
} as ComponentMeta<typeof AstroPopover>;

const Template: ComponentStory<typeof AstroPopover> = (args) => {
  const [isOpen, setIsOpen] = React.useState(true);

  return (
    <AstroPopover
      isOpen={isOpen}
      {...args}
      trigger={
        <AstroButton onClick={() => setIsOpen(!isOpen)}>Trigger</AstroButton>
      }
    />
  );
};

export const Default = Template.bind({});
Default.args = {
  text: "Popover",
};

export const TopPosition = Template.bind({});
TopPosition.args = {
  text: "Popover",
  position: "top",
};

export const BottomPosition = Template.bind({});
BottomPosition.args = {
  text: "Popover",
  position: "bottom",
};

export const TopPositionAndNubbin = Template.bind({});
TopPositionAndNubbin.args = {
  text: "Popover",
  position: "top",
  hasNubbin: true,
};

export const BottomPositionAndNubbin = Template.bind({});
BottomPositionAndNubbin.args = {
  text: "Popover",
  position: "bottom",
  hasNubbin: true,
};

export const LongText = Template.bind({});
LongText.args = {
  text: "I am a Popover with very long text!",
};

export const AllProps = Template.bind({});
AllProps.args = {
  text: "I am a Popover with very long text!",
  position: "top",
  hasNubbin: true,
}
