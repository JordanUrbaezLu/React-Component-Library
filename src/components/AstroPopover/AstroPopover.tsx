import * as React from "react";
import styles from "./AstroPopover.module.scss";
import classNames from "classnames";
import { getPopoverContainerHeight } from "./getPopoverContainerHeight";

export type Position = "bottom" | "left" | "right" | "top";

interface AstroPopoverProps {
  trigger: React.ReactNode;
  position?: Position;
  hasNubbin?: boolean;
  text: string;
  isOpen?: boolean;
}

export const AstroPopover: React.FC<AstroPopoverProps> = ({
  trigger,
  position = "bottom",
  hasNubbin = false,
  text,
  isOpen = false,
}) => {
  const [popoverHeight, setPopoverHeight] = React.useState<number>();

  const popoverContainerRef = React.useRef<HTMLDivElement>(null);
  const classes = classNames(
    styles.astroPopover,
    position === "bottom" && styles.bottom,
    position === "top" && styles.top,
    hasNubbin && styles.nubbin
  );

  React.useEffect(() => {
    const positionStyle = getPopoverContainerHeight({
      ref: popoverContainerRef,
    });
    setPopoverHeight(positionStyle);
  }, [isOpen]);

  return (
    <div ref={popoverContainerRef} className={styles.astroPopoverContainer}>
      {trigger}
      <div style={isOpen ? {transition: "opacity 0.5s", opacity: 1} : {opacity: 0}}>
        {isOpen && (
          <div
            style={
              position === "top"
                ? {
                    bottom: `${
                      (popoverHeight || 0) + (hasNubbin ? 7 : 0) + 4
                    }px`,
                  }
                : {}
            }
            className={classes}
          >
            {text}
          </div>
        )}
      </div>
    </div>
  );
};
