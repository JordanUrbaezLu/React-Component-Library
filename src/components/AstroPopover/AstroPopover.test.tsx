import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import AstroPopover, { AstroPopoverPosition } from "./AstroPopover";
import { axe, toHaveNoViolations } from "jest-axe";
expect.extend(toHaveNoViolations);

describe("AstroPopover", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <AstroPopover text="Popover" trigger={<button>Button</button>} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <AstroPopover text="Popover" trigger={<button>Button</button>} />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render text correctly.", () => {
    render(
      <AstroPopover
        isOpen={true}
        text="Popover"
        trigger={<button>Button</button>}
      />
    );
    expect(screen.getByText("Popover")).toBeInTheDocument();
  });
  test.each<AstroPopoverPosition>(["top", "bottom"])(
    "Should render positions correctly.",
    (position) => {
      render(
        <AstroPopover
          isOpen={true}
          position={position}
          text="Popover"
          trigger={<button>Button</button>}
        />
      );
      expect(screen.getByText("Popover")).toHaveClass(position);
    }
  );
  test("Should render trigger correctly.", () => {
    render(
      <AstroPopover
        isOpen={true}
        text="Popover"
        trigger={<button>Button</button>}
      />
    );
    expect(screen.getByRole("button", { name: "Button" })).toBeInTheDocument();
  });
  test("Should render hasNubbin correctly.", () => {
    render(
      <AstroPopover
        hasNubbin
        isOpen={true}
        text="Popover"
        trigger={<button>Button</button>}
      />
    );
    expect(screen.getByText("Popover")).toHaveClass("nubbin");
  });
  test("Should render isOpen correctly.", () => {
    render(
      <div data-testid="div">
        <AstroPopover
          isOpen={false}
          text="Popover"
          trigger={<button>Button</button>}
        />
      </div>
    );
    expect(screen.getByTestId("div")).not.toHaveTextContent("Popover");
  });
});
