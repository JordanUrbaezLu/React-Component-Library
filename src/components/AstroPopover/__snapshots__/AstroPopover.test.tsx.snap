// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstroPopover Should render component correctly. 1`] = `
<div
  class="astroPopoverContainer"
>
  <button>
    Button
  </button>
  <div
    style="opacity: 0;"
  />
</div>
`;
